components/blitz/src/ome/services/roi/GeomTool.java

```diff
@@ -354,6 +358,143 @@ public void handle(int x, int y) {
+           
+           // check if all shapes come from the same image
+           if (image == null) {
+               image = img;
+               pixels = image.getPrimaryPixels();
+           } else if (!image.getId().equals(img.getId()))
+               throw new ApiUsageException("All shapes have to be from the same image");
+           // check if z/t unattached fallback values are not out of bounds
+           if (zForUnattached < 0 || zForUnattached >= pixels.getSizeZ() ||
+               tForUnattached < 0 || tForUnattached >= pixels.getSizeT())
+               throw new ApiUsageException(
+                   "Fallback value(s) for unattached z/t shapes are out of bounds");
+           // if we have unattached z/t we use the unattached z,t fallback
+           final int theZ = shape.getTheZ() != null ? shape.getTheZ() : zForUnattached;
+           final int theT = shape.getTheT() != null ? shape.getTheT() : tForUnattached;
+           String lookupKey = theZ + "/" + theT;
+           List<ome.model.roi.Shape> lookupValue = zt_lookup.get(lookupKey);
+           if (lookupValue == null) {
+               lookupValue = new ArrayList<ome.model.roi.Shape>();
+               zt_lookup.put(lookupKey, lookupValue);
+           }
+           lookupValue.add(shape);
+       }
+
+       // any point iteration in a tiled image is a lost cause
+       if (data.requiresPixelsPyramid(pixels)) 
+           throw new ApiUsageException("This method can not handle tiled images yet.");
+       // check for channels filter
+       Set<Integer> validChannels = null;
+       if (channels != null && channels.length > 0) {
+           validChannels = new HashSet<Integer>(channels.length);
+           for (int ch : channels) {
+               if (ch < 0 || ch >= pixels.getSizeC())
+                   throw new ApiUsageException("Given channel(s) out of bounds.");
+               validChannels.add(ch);
+           }
+       }
+       // common info for all shapes 
+       final long pixelId = pixels.getId();
+       final int sizeX = pixels.getSizeX();
+       final int sizeY = pixels.getSizeY();
+       // loop over shapes (grouped by z/t planes)
+       for (final String key : zt_lookup.keySet()) {
+           final String[] keyTokens = key.split("/");
+           final int z = Integer.parseInt(keyTokens[0]);
+           final int t = Integer.parseInt(keyTokens[1]);
+
+           for (ome.model.roi.Shape shape : zt_lookup.get(key)) {
+               final SmartShape smartShape = (SmartShape) new ShapeMapper().map(shape);
+               final int size_stats = 
+                   validChannels != null ? validChannels.size() : pixels.getSizeC();
+               final ShapeStats stats = makeStats(size_stats);
+               stats.shapeId = shape.getId();
+               final double[] sumOfSquares = new double[size_stats];
+
+               final PixelBuffer buf = data.getBuffer(pixelId);
+               try {
```

Comment:
try with resources could be used to simplify the code

Reply:
unfortunately PixelBuffer's close can throw an IOException and that doesn't fare well with the try with resources paradigm. I'm afraid I'll need to catch this one since adding a throws to the method declaration to break the workflow in case of such an exception is not an option either...